Multiplayer Desync Debug Logging

AMZN-puvvadar opened the pull request:
Initial version of Desync Debug Logging accessible through the Debug Dropdown Menu (accessible with the Home key by default)
Includes:
- Logging of all desyncs
- Logging of inputs with non default values
- Logging of values specified by the user via macro
- ImGUI Dialog displaying logs with search and refresh functionality

The course of the work showed a general need for string converters for math types so these were moved into AzCore from EMFX. bool was specifically left alone as specifying it in AzCore seems to break AP.

Tests were not included for changes implemented in Multiplayer.Debug. Some tests were added for general MP changes. Linux has not been tested due to the lack of a machine.

Signed-off-by: puvvadar <puvvadar@amazon.com>

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -7,9 +7,177 @@
 {
+    enum class MultiplayerAuditCategory
+    {
+        MP_AUDIT_DESYNC,
```
Since this is an enum class, and you have to scope usage in any case, you can go with more readable names:

```
enum class AuditCategory
{
   Desync, 
   Input,
   Event
}
```

and then you have `Multiplayer::AuditCategory::Input` etc.. rather than triplicate the multiplayer bit, duplicate the audit bit, and use macro name formatting, `Multiplayer::MultiplayerAuditCategory::MP_AUDIT_INPUT`

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -7,9 +7,177 @@
+
+    //! @class IMultiplayerAuditingDatum
+    //! @brief IMultiplayerAuditingDatum provides an interface for auditing datums of a given auditing event
+    class IMultiplayerAuditingDatum
```
Try and move all your impls to an inline file, it's really difficult to see what your interfaces actually are (and there are no docs markup yet as well) with all the implementations mixed in together like this.

AMZN-puvvadar replied:
Good call, I've cleaned this up a fair bit.

lmbr-pip commented:
Added sig/core as PR involves new string conversions from EMFX to Code/Framework/AzCore/AzCore/Math/MathStringConversions.cpp files

lmbr-pip commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -7,9 +7,64 @@
+            return *this;
+        }
+
+        AZStd::string name;
```
Why break encapsulation by exposing member variables? These should also be m_ as per C++ coding standards (https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#variable-names)

AMZN-puvvadar replied:
This really should be a struct as it's an abstract value data container. Will fix the naming as well.

lmbr-pip commented on Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerDebug.h:
```diff
@@ -0,0 +1,85 @@
+    template<class T>
+    class MultiplayerAuditingDatum : public IMultiplayerAuditingDatum
+    {
+    private:
```
nit: Consider public things first rather than private. As a user I care about what I can call, not so much the implementation

lmbr-pip commented on Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:
```diff
@@ -189,15 +217,24 @@ namespace Multiplayer
 
                 if (lostInput)
                 {
+                    //TODO: Maybe add to Audit Trail here (server)
```
Consider migrating TODOs to public GitHub issues. TODOs are hard for other developers to take action on as its unclear who is doing the work or why

lmbr-pip commented on Gems/Multiplayer/Code/Source/Components/LocalPredictionPlayerInputComponent.cpp:
```diff
@@ -312,30 +352,39 @@ namespace Multiplayer
 
-#ifndef AZ_RELEASE_BUILD
+        const uint32_t inputHistorySize = static_cast<uint32_t>(m_inputHistory.Size());
+        const uint32_t historicalDelta = aznumeric_cast<uint32_t>(m_clientInputId - inputId); // Do not replay the move we just corrected, that was already processed by the server
```
nit: avoid personal  pronouns in comments 'we' is often unclear to the reader. Also as this code can be run in game, server or editor again, the we is unclear.

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:
```diff
@@ -0,0 +1,72 @@
+
+        AZStd::string_view m_filter;
+        AzFramework::DebugDisplayRequests* m_debugDisplay = nullptr;
+        char m_inputBuffer[1028] = {};
```
Consider moving magic number to const or similar and include how set. Also note fixed sized 'dumb' arrays have historically caused a lot of pain points around  overflow so be careful that use of buffer is appropriately guarded.

AMZN-puvvadar replied:
From what I'm understanding of ImGui (the user of this buffer), it should be enforced. I'll add a const.

lmbr-pip reviewed (commented):
Can we update description as it seems out of date?

Also be add:
* How tested?
* If you tested on Linux, as its now reference platform?
* Reason for no automated tests (unit, integration) with code.

lmbr-pip commented on Code/Framework/AzCore/AzCore/Math/MathStringConversions.h:
```diff
@@ -0,0 +1,51 @@
+namespace AZStd
+{
+    //! Prints a Vector2 with precision to 8 decimal points
+    void to_string(string& str, const AZ::Vector2& value);
```
I believe these names should be camel case as per core coding standards: https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md

AMZN-puvvadar replied:
Which names? The variables are technically camel case by virtue of being single words and to_string while not, is the existing standard for the API.

kberg-amzn replied:
AZStd follows the std naming conventions in all places for public methods. Hence Rajiv has picked the correct method names in this case.

lmbr-pip commented on Gems/EMotionFX/Code/MCore/Source/StringConversions.h:
```diff
@@ -50,18 +41,6 @@ namespace MCore
 namespace AZStd
 {
     void to_string(string& str, bool value);
```
nit: Why not lift all the conversions out?

AMZN-puvvadar replied:
The bool conversion specifically breaks AssetProcessor for reasons that are beyond me. I'm consequently keeping it scoped.

burelc-amzn replied:
The bool conversion is gross. It adds an overload that has significant impact on overload resolution. Consider these
overloads:

```cpp
void to_string(AZStd::string& dest, AZStd::wstring_view src);
void to_string(string& str, bool value);
```

And then calling code like this:
```
WCHAR src[260];
AZStd::string dst;
AZStd::to_string(dst, src); // Which overload does this call?
```

If the .cpp has not included `MCore/Source/StringConversions.h`, the call
to `to_string()` will convert the `WCHAR[260]` type to a
`AZStd::wstring_view`, and call the first overload. But if
`StringConversions.h` _has_ been included, the implicit conversion of
`WCHAR[260]` to `bool` has a higher precedence, and it will be chosen
instead.

This overload was causing some uses of `to_string` in
`AnimGraph/GameController.cpp` to resolve to the wrong overload in unity
builds. I have a commit that I made this morning to remove it completely.

lumberyard-employee-dm replied:
We can move this to AzCore/Math as well.
We just need to use a constraint that the `bool` overload accepts a type that is exactly `bool` once cv ref qualifiers have been removed
https://godbolt.org/z/G3ce383T1
```
template<class BoolType>
auto to_string(std::string& dest, BoolType value) ->
    std::enable_if_t<std::same_as<std::remove_cvref_t<BoolType>, bool>>
```

AMZN-puvvadar replied:
I'm going to pull the to_string changes out into a separate PR. Now that I'm aware of AzCore/Math/ToString, I think it's warranted.

lmbr-pip commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -7,9 +7,70 @@
+    {
+    public:
+        MultiplayerAuditingElement() = default;
+        MultiplayerAuditingElement(const MultiplayerAuditingElement& rhs)
```
Did we consider move assignment operator for struct to avoid deep clones? or is there always a reason for deep clones?

AMZN-puvvadar replied:
The usage of unique_ptr in the elements vector (which allows it to function as an abstract value vector) forces it.

lmbr-pip commented on Gems/Multiplayer/Code/Include/Multiplayer/NetworkInput/NetworkInput.h:
```diff
@@ -52,6 +53,11 @@ namespace Multiplayer
 
         bool Serialize(AzNetworking::ISerializer& serializer);
 
+        void LogComponentInputDelta() const;
```
nit: consider autogen comments on all public headers

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:
```diff
@@ -0,0 +1,75 @@
+        AZStd::string_view GetAuditTrialFilter();
+
+        //! Sets string filter for the audit trail
+        void SetAuditTrailFilter(AZStd::string_view filter);
```
nit: pass as const ref?

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:
```diff
@@ -0,0 +1,75 @@
+        ~MultiplayerDebugAuditTrail();
+
+        //! Main update loop.
+        void OnImGuiUpdate(AZStd::deque<AuditTrailInput> auditTrailElems);
```
nit: Why pass by value?

AMZN-puvvadar replied:
Looks like an oversight

lmbr-pip commented on Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp:
```diff
@@ -86,6 +86,39 @@ namespace Multiplayer
+    {
+        for (uint16_t i = 0; i < m_componentInputs.size(); ++i)
+        {
+            //AZLOG_WARN(m_componentInputs[i].get()->GetInputDeltaLog().c_str());
```
minor: Avoid committing commented out code. Consider if function has any value as is

lmbr-pip commented on Gems/Multiplayer/Code/Source/NetworkInput/NetworkInputChild.cpp:
```diff
@@ -25,6 +25,16 @@ namespace Multiplayer
         return *this;
     }
 
+    bool NetworkInputChild::operator== (const NetworkInputChild& rhs) const
+    {
+        return m_owner == rhs.m_owner && m_networkInput.m_inputId == rhs.m_networkInput.m_inputId;
+    }
+
+    bool NetworkInputChild::operator!= (const NetworkInputChild& rhs) const
+    {
+        return !(m_owner == rhs.m_owner);
+    }
```
Any updates on this? Is operator == still the right thing todo here?

AMZN-puvvadar replied:
I believe so, though I'd welcome feedback from other sig-network contributors.

kberg-amzn replied:
Yeah, this is fine.. it's required only for the delta tracking functionality so we can delta compress state updates through the serializer.

lmbr-pip reviewed (commented):
LG.

Pretty large PR. I would consider potentially breaking work up like this into smaller chunks in future, as theres a lot of details here.

kberg-amzn commented on Code/Framework/AzCore/AzCore/Math/MathStringConversions.cpp:
```diff
@@ -0,0 +1,71 @@
+/*
```
nice!

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -7,9 +7,70 @@
 namespace Multiplayer
 {
+    //! Categories for Auditing logs used in MultiplayerDebugAuditTrail
+    enum class MultiplayerAuditCategory
```
You're still duping Multiplayer here, although it's much less of a big deal than before.. ie `Multiplayer::MultiplayerAuditCategory::Desync` instead of simply `Multiplayer::AuditCategory::Desync`

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -7,9 +7,70 @@
+        MultiplayerAuditingElement(const MultiplayerAuditingElement& rhs)
+        {
+            m_name = rhs.m_name;
+            m_elements.resize(rhs.m_elements.size());
```
resize_no_construct?

AMZN-puvvadar replied:
Oh good call!

AMZN-puvvadar replied:
I had to actually stick to resize. unique_ptr ends up attempting to invoke deleter on assignment so the lvalue seems to need to be constructed.

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -24,5 +85,36 @@ namespace Multiplayer
+            AZStd::vector<MultiplayerAuditingElement>&& entryDetails) = 0;
     };
+
+    #define AZ_MPAUDIT_INPUT_REWINDABLE(INPUT, REWINDABLE, VALUE_TYPE)                                                            \
```
I would comment up these macros so everyone knows exactly what they do and what the parameters are.

kberg-amzn commented on Gems/Multiplayer/Code/Include/Multiplayer/NetworkTime/RewindableObject.h:
```diff
@@ -64,6 +64,10 @@ namespace Multiplayer
 
+        //! Const base type retriever for value at last known authoritative update or the nearest to it.
+        //! @return value in const base type form
+        const BASE_TYPE& GetAuthority() const;
```
This guy is a bit concerning. I think it might be worth having another look at. m_authTime is only updated on a serialize call, which is a bit undefined to be honest (could be a correction, could be a property update, etc..) and is zero at all other times. Also I don't know what the behaviour should be on a client vs a server.

Coupling to a role like authority at this level seems a bit wrong. You might want to back this up to 'GetLastSerializedValue()' or something a bit more generalized.

AMZN-puvvadar replied:
I'll take a deeper look through this.

kberg-amzn reviewed (approved):
Geez you did a lot of good cleanup work on this. A few more comments, especially around the RewindableObject change that I really think should get looked at more carefully before submit, but overall really good job :thumbsup:

AMZN-puvvadar commented:
Latest changes addressed some smaller feedback points and moves Auditing Macros from IMultiplayerDebug.h to MultiplayerDebug.h. This is to simplify the usage requirements of the macros as the template specifications in MultiplayerDebug.h are required to use the macros anyway.

AMZN-puvvadar commented:
Most recent change changes the authority value logic added to RewindableObject to last serialized value and adds a couple of guards around it. It also fixes a variety of exceptions (some of which that were in Debug mode only).

kberg-amzn commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugSystemComponent.cpp:
```diff
@@ -499,8 +542,145 @@ namespace Multiplayer
+                            if (filteredChild.m_elements.size() > 0)
+                            {
+
+                            filteredInput.m_children.push_back(filteredChild);
```
There's a bunch of weird formatting going on throughout these changes.. I'm not entirely sure what happened, but it's worth a quick pass to clean up anything that got a bit munged..

AMZN-puvvadar replied:
Hmm yeah, this function has some oddities. I've cleaned them up.

AMZN-puvvadar commented:
Fixed another IMGUI assert and cleaned up cvar usage. In particular, moved the Input auditing flag to LocalPredictionPlayerInputComponent where it more efficiently gates logic and got rid of the Event auditing flag entirely since it's an opt in feature.

AMZN-puvvadar commented:
Latest set of changes adds a few things

- AZ_MPAUDIT_VALUE macro to let developers audit a variable when a NetworkInput isn't readily accessible, useful when several calls deep
- Updated Refresh logic for more consistency
- All Multiplayer Debug popups no longer let mouse clicks pass through to the game. In the Audit Trail in particular, this prevents event pollution when trying to refresh
- Added (limited) Desync events on the Server so the Audit Trail populates. This is useful when needing to compare Values across endpoints

lsemp3d reviewed (commented):
Where these functions unused in EmotionFX? I don't see any updates to the EMFX codebase to use the new AzCore include path and namespace

AMZN-puvvadar commented:
> Were these functions unused in EmotionFX? I don't see any updates to the EMFX codebase to use the new AzCore include path and namespace

The to_string functions? EMFX was namespacing them under AZStd so the change to include MathStringConversion.h cleanly replaced any usages.

AMZN-puvvadar commented:
to_string changes are now represented in https://github.com/o3de/o3de/pull/7609 

If and when that change is approved and merged, I'll pull it here to remove those changes from this review. Thanks for the input everyone.

lumberyard-employee-dm commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:
```diff
@@ -37,7 +37,7 @@ namespace Multiplayer
+    void MultiplayerDebugAuditTrail::SetAuditTrailFilter(const AZStd::string filter)
     {
         m_filter = filter;
     }
```
First this parameter shouldn't be const.

Second the filter string should be moved into m_filter.

By making the parameter const it inhibits moving of the filter parameter into m_filter
```suggestion
    void MultiplayerDebugAuditTrail::SetAuditTrailFilter(AZStd::string filter)
    {
        m_filter = AZStd::move(filter);
    }
```

AMZN-puvvadar replied:
I hadn't thought of move here, I'll make that change.

AMZN-puvvadar commented:
Changes made in https://github.com/o3de/o3de/pull/7609 have been merged and resolved

lmbr-pip commented on Gems/EMotionFX/Code/MCore/Source/StringConversions.h:
```diff
@@ -35,4 +35,4 @@ namespace MCore
     };
-}
\ No newline at end of file
+}
```
nit: appears to be just Whitespace change. Consider reverting

lmbr-pip commented on Gems/Multiplayer/Code/Include/Multiplayer/IMultiplayerDebug.h:
```diff
@@ -7,9 +7,70 @@
+    };
+
+    //! @class IMultiplayerAuditingDatum
+    //! @brief IMultiplayerAuditingDatum provides an interface for auditing datums of a given auditing event
```
nit: "provides an interface for auditing datums of a given auditing event" seems like too many uses of word "auditing"  - it may be accurate, worth a quick review if this is really the responsibilities of this class.

AMZN-Olex reviewed (approved):
Looks good!

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:
```diff
@@ -0,0 +1,251 @@
+        m_updateDebugOverlay.Enqueue(AZ::TimeMs{ 0 }, true);
+    }
+
+    MultiplayerDebugAuditTrail::~MultiplayerDebugAuditTrail()
```
nit: Move to default destructor?

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:
```diff
@@ -0,0 +1,251 @@
+        m_filter = AZStd::move(filter);
+    }
+
+    // --------------------------------------------------------------------------------------------
```
nit: Not sure why the horizontal line comment here? What are we conveying?

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.cpp:
```diff
@@ -0,0 +1,251 @@
+
+    bool MultiplayerDebugAuditTrail::CanPumpAuditTrail()
+    {
+        bool canPump = m_canPumpTrail;
```
nit: Not too sure why this changes state. Can<I do something> seems like it should not change internal state.

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:
```diff
@@ -0,0 +1,75 @@
+{
+    struct AuditTrailInput
+    {
+        AuditTrailInput(AuditCategory category, ClientInputId inputId, HostFrameId hostFrameId,
```
nit: should any of these be const ref, rather than passed by value?

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:
```diff
@@ -0,0 +1,75 @@
+    };
+
+    //! Buffer size for ImGui Search bar in Audit Trail UI
+    const int AUDIT_SEARCH_BUFFER_SIZE = 1024;
```
nit: Considering adding comments about how set, or implications for changing value.

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugAuditTrail.h:
```diff
@@ -0,0 +1,75 @@
+        bool CanPumpAuditTrail();
+
+        //! Gets string filter for the audit trail
+        AZStd::string_view GetAuditTrialFilter();
```
nit: Should any of these be const methods?

lmbr-pip commented on Gems/Multiplayer/Code/Source/Debug/MultiplayerDebugSystemComponent.cpp:
```diff
@@ -440,67 +477,238 @@ namespace Multiplayer
+
+        for (auto elem = m_committedAuditTrail.begin(); elem != m_committedAuditTrail.end(); ++elem)
+        {
+            const char* nodeTitle = elem->m_category == AuditCategory::Desync
```
Nit: consider using params make nested ternarys easier to read / or split up to aid maintainers.

lmbr-pip commented on Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp:
```diff
@@ -86,6 +86,31 @@ namespace Multiplayer
+    AZStd::vector<MultiplayerAuditingElement> NetworkInput::GetComponentInputDeltaLogs() const
+    {
+        AZStd::vector<MultiplayerAuditingElement> logs;
+        for (uint16_t i = 0; i < m_componentInputs.size(); ++i)
```
nit: avoid 'i' as var name

lmbr-pip commented on Gems/Multiplayer/Code/Source/NetworkInput/NetworkInput.cpp:
```diff
@@ -86,6 +86,31 @@ namespace Multiplayer
+        {
+            return owner->GetName();
+        }
+        return "null owner";
```
nit: Reason for a specific string rather than empty string?

lmbr-pip reviewed (commented):
Have some structs and classes with public member variables, consider if you really need to break encapsulation here.

Have a number of class methods that look like they could be const methods in interfaces.

Looks good overall. Some minor comments, leave to your discretion about answering.
